### ISSUE TITLE
`helm package` should produce bitwise-deterministic tar files

### ISSUE THREAD

**Issue by cmaher** [helm/helm#3612]
We have a CI process for publishing charts where we try to determine if pushed repository needs to have it's charts built and published. We would like to do this with a bitwise comparison of the generated tgz files. This works for charts without dependencies, but it is inconsistent for charts with dependencies.

For the setup (Using v2.8.1):
```
charts
├── child1
│   └── Chart.yaml
├── child2
│   └── Chart.yaml
├── child3
│   └── Chart.yaml
└── parent
    ├── Chart.yaml
    └── requirements.yaml
```

With charts:
```
# child{1..3}/Chart.yaml
name: child{1..3}
version: 1.0.0

# parent/Chart.yaml
name: parent
version: 1.0.0

# parent/requirements.yaml
dependencies:
- name: child1
  repository: file://../child1
  version: "*"
- name: child2
  repository: file://../child2
  version: "*"
- name: child3
  repository: file://../child3
  version: "*"

```

Running:
```
# working dir: parent
$ helm dep update
$ for i in {1..1000}; do helm package . && sha256sum parent-1.0.0.tgz && rm parent-1.0.0.tgz; done
```
produces several (3, after a significant number of iterations) different shas. The shas also appear to be non-uniformly distributed (e.g. 1 sha appears 70% of the time).

My guess is that https://github.com/kubernetes/helm/blob/master/pkg/chartutil/save.go#L160 is iterating over dependencies in a non-deterministic order, thus producing different tar files.

**Comment by bacongobbler** [comment]
I think this is actually due to gzip. Part of the gzip header has a mod time for whatever is compressed in the file, so by calling `helm package` multiple times, each one creates their own tarball which is then packaged by gzip and given a unique shasum because the modtime differed.

If you `gunzip` the .tgz, is the shasum identical between tarballs?

**Comment by cmaher** [comment]
`gunzip`ing the tars still produces different tarballs. There still appear to be 3 different shasums.

**Comment by bacongobbler** [comment]
Following up on my previous comment, you can even reproduce this yourself at home:

```
><> helm fetch stable/traefik
><> gunzip traefik-1.24.1.tgz 
><> helm fetch stable/traefik
><> shasum -a 256 *
a032f1eba2c54b76d8de38d3763bab5ca5a6383af1156f30caf4595adaec67a0  traefik-1.24.1.tar
ebd78bdbbf63f557db8646c6351bfeeac849951334d4cbfc103ecc6bb1b170c5  traefik-1.24.1.tgz
```

So in this example, I've shown that I've got stable/traefik, both in compressed and decompressed versions from the official stable/traefik chart. If I gzip this up, we should be able to assert that we come up with the same shasum because the contents are the same, right?

Wrongo.

```
><> gzip traefik-1.24.1.tar 
><> shasum -a 256 *
144eaa108aff820dd754e16732bbacd29accb41954d6249b4111623587fae415  traefik-1.24.1.tar.gz
ebd78bdbbf63f557db8646c6351bfeeac849951334d4cbfc103ecc6bb1b170c5  traefik-1.24.1.tgz
```

But the shasums of both the recently-gzipped tarball and the old shasum are still identical.

```
><> gunzip traefik-1.24.1.tar.gz 
><> mv traefik-1.24.1.tar traefik-1.24.1.tar.1
><> gunzip traefik-1.24.1.tgz 
><> mv traefik-1.24.1.tar traefik-1.24.1.tar.2
><> shasum -a 256 *
a032f1eba2c54b76d8de38d3763bab5ca5a6383af1156f30caf4595adaec67a0  traefik-1.24.1.tar.1
a032f1eba2c54b76d8de38d3763bab5ca5a6383af1156f30caf4595adaec67a0  traefik-1.24.1.tar.2
```

So in other words, if you want a deterministic shasum of the contents of the chart, I'd suggest decompressing the packages first before checking the shasum.

Would that work for you?

**Comment by bmperrea** [comment]
The timestamp issue in the default gzip settings appears to be irrelevant since there are only a few unique shas that appear in thousands of tests on `helm package` (whereas the shas from the gzip example appear to be purely random). It looks like `writeToTar` in `pkg/chartutil/save.go` specifies its own header here:  https://github.com/kubernetes/helm/blob/master/pkg/chartutil/save.go#L205.

Perhaps sorting the dependencies in a deterministic manner before serialization is all that is needed, but any help on where to put the sort would be appreciated.

For now a workaround is to do a file diff on the decompressed packages after removing the dependency tars. This won't catch differences in dependencies except version differences (which can happen for local dependencies, e.g.), but will catch most differences until this can be patched.

**Comment by bacongobbler** [comment]
I was unable to reproduce this behaviour using a stable chart. Can you provide a sample chart that produces this behaviour?

My steps to test this:

```
$ helm fetch --untar stable/traefik
$ for i in {1..1000}; do helm package traefik/ && mv traefik-1.24.1.tgz traefik-1.24.1.tgz.$i; done
$ shasum -a 256 *.tgz* | awk '{print $1}' | uniq
ebd78bdbbf63f557db8646c6351bfeeac849951334d4cbfc103ecc6bb1b170c5
```

Every invocation of `helm package` returned a consistent shasum.

/shrug

**Comment by cmaher** [comment]
https://github.com/cmaher/nondeterministic-charts demonstrates this using `Client: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}`

**Comment by jeff-knurek** [comment]
I found out a way to reproduce this. One thing that seems to be helpful in reproducing is having more than one dependency in the requirements file. Also, the large the values.yaml file  the easier it is to reproduce as well. 

`spinnaker` is an example stable chart to reproduce:
```
mkdir pkgs
cd spinnaker
helm dep up
for i in {1..30}; do helm package . && mv spinnaker-0.5.0.tgz ../pkgs/spinnaker-0.5.0.tgz.$i; done
cd ../pkgs
shasum -a 256 *.tgz* | awk '{print $1}' | uniq
```

**Comment by diclophis** [comment]
We have been able to reproduce this issue with some of our local charts as well.

**Comment by adrian-gierakowski** [comment]
The problem definitely manifests itself on macOS. None of the suggestions mentioned in this tread solve the issues. I've created a repo to demonstrate the problem and a workaround: https://github.com/adrian-gierakowski/helm-package-sha-test.

**Comment by jdolitsky** [comment]
Is this related to dependencies, or is the issue just increasingly more likely to occur the more files a chart contains?

> The issue is that the POSIX directory enumeration API does not define a specific order for file names to be returned. There may be a "common" order but since sorting/ordering is not something that can be done generically (it is locale/user specific), the core filesystems are not forced to provide a consistent order. Result orders may depend on available cache entries or order of data returned from elevator algorithms that read from the disk or a number of other out-of-order execution behaviors.

Is there really not a way enforce this order in Go? @Michael-Sinz

**Comment by Michael-Sinz** [comment]
You can write your own code to take the enumeration and sort it.  Pick your sort to be consistent across all locales.

Unix/Posix does not push costly operations onto all use cases when most cases do not need that - it assumes that you would compose one operation with another to get the desired result.

__
Michael Sinz C Architect C <http://midori/> v C Microsoft
________________________________

**Comment by Michael-Sinz** [comment]
Is there no progress on this issue.  The lack of deterministic output for constant input is a big problem.  It means that helm package is not deterministic and we can not depend on it producing the same results.  Imaging if a compiler could make different code for the same source code from run to run - how would you debug such a program?

**Comment by rjeberhard** [comment]
Agreed. We've had to make clumsy workarounds to only package the chart if one of the source files changes. One issue is the embedded time stamps.

**Comment by bacongobbler** [comment]
@Michael-Sinz I have not seen any activity from the community on this bug. Please feel free to work on a fix!

**Comment by Michael-Sinz** [comment]
We still would like to have deterministic helm chart building.  The fact that the same helm chart from the same source tree from the same git hash with the same helm version does not produce deterministic hashes is concerning.

**Comment by sathieu** [comment]
Deterministic output is the goal of the [Reproducible builds ](https://reproducible-builds.org/) project.

It has a recommended [tar example](https://reproducible-builds.org/docs/archives/#full-example):

```shell
# requires GNU Tar 1.28+
$ tar --sort=name \
      --mtime="@${SOURCE_DATE_EPOCH}" \
      --owner=0 --group=0 --numeric-owner \
      --pax-option=exthdr.name=%d/PaxHeaders/%f,delete=atime,delete=ctime \
      -cf product.tar build
```

`gz` seems to produce reproducible output [since 2018](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=32342).


But as `helm` uses go libs, maybe the fix is around file order:

https://github.com/helm/helm/blob/f41f46c6c6a1fae459b51a110f76d60f33281f74/pkg/chartutil/save.go#L205-L228

**Comment by geowalrus4gh** [comment]
Hi, we are really waiting for this issue to be fixed. Since real advantage of using OCI is that we make use of the delta changes in the chart that are only committed to the repository and this saves the space at the repository. We are leveraging this feature in OCI container images a lot. Without this feature, we cannot declare helm OCI support is moved to permanent from experimental.

**Comment by dinvlad** [comment]
Any updates?

**Comment by h0tbird** [comment]
I can reproduce it like this:
```console
$ helm version
version.BuildInfo{Version:"v3.10.0", GitCommit:"ce66412a723e4d89555dc67217607c6579ffcb21", GitTreeState:"clean", GoVersion:"go1.19.1"}
$ helm create foo
Creating foo
$ for i in {1..10}; do helm package foo >/dev/null; md5sum foo-0.1.0.tgz; sleep 0.5; done
39e2b6bff940ed840873926f271d72f8  foo-0.1.0.tgz
f34d771d8039d77c9b76340c7940ab55  foo-0.1.0.tgz
f34d771d8039d77c9b76340c7940ab55  foo-0.1.0.tgz
afe15ee8c34f069e967e88e70c7bbd29  foo-0.1.0.tgz
afe15ee8c34f069e967e88e70c7bbd29  foo-0.1.0.tgz
7c1b20acd7a9a73944af13fec2303cf0  foo-0.1.0.tgz
7c1b20acd7a9a73944af13fec2303cf0  foo-0.1.0.tgz
16cef47516265bed7f8dbd9ef145042c  foo-0.1.0.tgz
16cef47516265bed7f8dbd9ef145042c  foo-0.1.0.tgz
1532cb3b9a5d828e19d74b034caebace  foo-0.1.0.tgz
```

**Comment by dinvlad** [comment]
FWIW the example from [reproducible builds](https://github.com/helm/helm/issues/3612#issuecomment-947237805) worked well for me, but that obviously requires custom scripting, which is not ideal.

**Comment by FrenchBen** [comment]
Fix for the above has been pending review/merge. Not sure what the blocker is TBH: 
https://github.com/helm/helm/pull/10178
